File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    List<String> ans = new LinkedList<String>();
    String[] mp = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    private void foo (int i, String temp, String digits) {
        if(i == digits.length()){
            if(temp.length() > 0) ans.add(temp);
            return;
        }
        for(int k=0; k<mp[digits.charAt(i) - '0'].length(); k++)
            foo(i+1, temp + mp[digits.charAt(i) - '0'].charAt(k), digits);
    }

    public List<String> letterCombinations(String digits) {
        foo(0, "", digits);
        return ans;
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> ans;
    string mp[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    string temp = "";

    void dfs(int i, string digits){
        if(i == digits.size()){
            if(!temp.empty()) ans.push_back(temp);
            return;
        }
        for(char c : mp[digits[i] - '0']){
            temp += c;
            dfs(i+1, digits);
            temp.pop_back();
        }
    }

    vector<string> letterCombinations(string digits) {
        dfs(0, digits);
        return ans;
    }
};







        // dfs(0, digits);
    // void dfs (int i, string digits) {
    //     if(i == digits.size()){
    //         if(!temp.empty()) ans.push_back(temp);
    //         return;
    //     } 
    //     for(char c : map[digits[i] - '0']){
    //         temp.push_back(c);
    //         dfs(i+1, digits);
    //         temp.pop_back();
    //     }
    // }